File: backend/src/task_center/context_engine/scope.py
"""ContextScope — identity surface every recipe sees.

The scope carries identity (goal / iteration / attempt / task ids). It
does **not** carry store handles; those live on :class:`ContextEngineDeps`
so recipes can be swapped without touching call sites.

The role-specific factory classmethods (:meth:`for_planner`,
:meth:`for_generator`, etc.) document the required fields per role at the
call site: omitting one raises ``TypeError`` at call time, and strict
mypy will narrow the kwargs to their declared ``str`` types. The engine
still validates via :meth:`assert_fields` so direct ``ContextScope(...)``
construction is also covered at runtime.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from task_center.context_engine.exceptions import RecipeScopeError

ScopeField = Literal["workflow_id", "iteration_id", "attempt_id", "task_id"]


@dataclass(frozen=True, slots=True)
class ContextScope:
    """Identity surface threaded through resolver + engine + recipes."""

    workflow_id: str | None = None

    # Optional identity fields — recipes declare which of these they need.
    iteration_id: str | None = None
    attempt_id: str | None = None
    task_id: str | None = None

    def assert_fields(self, required: frozenset[str]) -> None:
        """Raise :class:`RecipeScopeError` if any required field is None."""
        missing = sorted(f for f in required if getattr(self, f, None) is None)
        if missing:
            raise RecipeScopeError(f"ContextScope is missing required fields: {missing!r}")

    def require_field(self, field: ScopeField) -> str:
        """Return one required identity field as a non-optional string."""
        value = getattr(self, field)
        if not isinstance(value, str):
            raise RecipeScopeError(f"ContextScope is missing required field: {field!r}")
        return value

    # ---- Role-specific factory shortcuts -------------------------------
    #
    # Each factory takes ONLY the required fields for that recipe role as
    # positional/keyword args. Missing a required field is a static error
    # instead of a runtime assert. The defaults flow through the dataclass
    # for any optional fields the role might inspect.

    @classmethod
    def for_planner(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
    ) -> ContextScope:
        """Scope shape required by the planner recipe."""
        return cls(
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
        )

    @classmethod
    def for_generator(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
        task_id: str,
    ) -> ContextScope:
        """Scope shape required by the generator recipe."""
        return cls(
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
            task_id=task_id,
        )

    @classmethod
    def for_evaluator(
        cls,
        *,
        workflow_id: str,
        iteration_id: str,
        attempt_id: str,
    ) -> ContextScope:
        """Scope shape required by the evaluator recipe."""
        return cls(
            workflow_id=workflow_id,
            iteration_id=iteration_id,
            attempt_id=attempt_id,
        )

File: agent-core/crates/eos-workflow/src/context/scope.rs
use eos_types::{AttemptId, IterationId, TaskId, WorkItemId, WorkflowId};

use super::ContextRole;

/// Role-specific launch scope. Each role carries only the ids it needs.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ContextScope {
    /// Planner launch context.
    Planner {
        /// Workflow id.
        workflow_id: WorkflowId,
        /// Iteration id.
        iteration_id: IterationId,
        /// Attempt id.
        attempt_id: AttemptId,
        /// Opaque planner task id.
        task_id: TaskId,
    },
    /// Worker launch context.
    Worker {
        /// Workflow id.
        workflow_id: WorkflowId,
        /// Iteration id.
        iteration_id: IterationId,
        /// Attempt id.
        attempt_id: AttemptId,
        /// Opaque worker task id.
        task_id: TaskId,
        /// Planner-authored work item id.
        work_item_id: WorkItemId,
    },
}

impl ContextScope {
    /// Scope for a planner launch.
    #[must_use]
    pub fn for_planner(
        workflow_id: WorkflowId,
        iteration_id: IterationId,
        attempt_id: AttemptId,
        task_id: TaskId,
    ) -> Self {
        Self::Planner {
            workflow_id,
            iteration_id,
            attempt_id,
            task_id,
        }
    }

    /// Scope for a worker launch.
    #[must_use]
    pub fn for_worker(
        workflow_id: WorkflowId,
        iteration_id: IterationId,
        attempt_id: AttemptId,
        task_id: TaskId,
        work_item_id: WorkItemId,
    ) -> Self {
        Self::Worker {
            workflow_id,
            iteration_id,
            attempt_id,
            task_id,
            work_item_id,
        }
    }

    /// The launch role this scope was built for.
    #[must_use]
    pub const fn role(&self) -> ContextRole {
        match self {
            Self::Planner { .. } => ContextRole::Planner,
            Self::Worker { .. } => ContextRole::Worker,
        }
    }
}
